File: src/ee/Notification.cpp
//
//  NotificationProtocol.cpp
//  ee_core
//
//  Created by Zinge on 3/28/17.
//
//

#include "ee/Notification.hpp"
#include "ee/NotificationBuilder.hpp"
#include "ee/internal/MessageBridge.hpp"
#include "ee/libs/nlohmann/json.hpp"

namespace ee {
void Notification::schedule(const NotificationBuilder& builder) {
    nlohmann::json json;
    json["title"] = builder.title_;
    json["body"] = builder.body_;
    json["delay"] = builder.delay_;
    json["interval"] = builder.interval_;
    json["tag"] = builder.tag_;
    MessageBridge::getInstance().call("__notification_schedule", json.dump());
}

void Notification::unschedule(int tag) {
    nlohmann::json json;
    json["tag"] = tag;
    MessageBridge::getInstance().call("__notification_unschedule", json.dump());
}

void Notification::unscheduleAll() {
    MessageBridge::getInstance().call("__notification_unschedule_all");
}
} // namespace ee

File: src/ts/src/notification/internal/Notification.ts
import {
    ILogger,
    IMessageBridge,
    Platform,
} from "../../core";
import { INotification } from "../INotification";
import { NotificationBuilder } from "../NotificationBuilder";

type Destroyer = () => void;

export class Notification implements INotification {
    private readonly kPrefix = `NotificationBridge`;
    private readonly kSchedule = `${this.kPrefix}Schedule`;
    private readonly kUnschedule = `${this.kPrefix}Unschedule`;
    private readonly kClearAll = `${this.kPrefix}ClearAll`;

    private readonly _bridge: IMessageBridge;
    private readonly _logger: ILogger;
    private readonly _destroyer: Destroyer;

    public constructor(bridge: IMessageBridge, logger: ILogger, destroyer: Destroyer) {
        this._bridge = bridge;
        this._logger = logger;
        this._destroyer = destroyer;
    }

    public destroy(): void {
        this._destroyer();
    }

    public schedule(builder: NotificationBuilder): void {
        const request = {
            ticker: builder.ticker,
            title: builder.title.length === 0 ? Platform.getApplicationName() : builder.title,
            body: builder.body,
            delay: builder.delay,
            interval: builder.interval,
            tag: builder.tag,
        };
        this._bridge.call(this.kSchedule, JSON.stringify(request));
    }

    public unschedule(tag: number): void {
        const request = {
            tag: tag,
        };
        this._bridge.call(this.kUnschedule, JSON.stringify(request));
    }

    public clearAll(): void {
        this._bridge.call(this.kClearAll);
    }
}